### DISCUSSION TITLE
Chat backup feature

### DISCUSSION THREAD

**Discussion by mike-lloyd03** [BlueBubblesApp/bluebubbles-app#2750] [Ideas]
Currently, OpenBubbles has a feature to backup and restore a user's chat history from the filesystem. Is there any interest in having this feature in BlueBubbles? I'd like to open a PR which just ports over the changes from `lib/app/layouts/settings/pages/server/backup_restore_panel.dart` that @TaeHagen already added to the OpenBubbles fork (a few other minor changes would be needed as well).

My use case for this is moving chat history from BlueBubbles to OpenBubbles. But I'm sure there are other uses for this feature. I'm not sure how useful the restore functionality would be since BB can sync message history with the server, but I suppose it might be helpful if both the mobile app and server lost chat history for some reason.

**Comment by zlshames** [comment]
Do you have a link to Tae's code? I would like to see how he does it. Sorry for the late reply.

**Comment by mike-lloyd03** [comment]
Yeah. Here's the commit where he added the logic:
https://github.com/TaeHagen/bluebubbles-app/commit/2e9c06af3ca144f589655db01ea9e216ab4daef8

**Comment by zlshames** [comment]
hmmm ok, so it looks like he just pulls the chats & messages and then JSON encodes them and stores them in files. I'm not sure I love that idea tbh... If i were to do it, i'd do it in a different manner. It also base64 encodes any attachments and stores it in the JSON as well. What I'm worried about is the JSON files becoming gigabytes or even a few hundred MB.

**Comment by TaeHagen** [comment]
So I actually changed that because it was causing issues with OOM errors with 500mb+ files. I changed it to a very basic but functional serial format with json for messages and attachments in binary after the json. Look at the latest commit.

**Reply by zlshames** [reply]
that's what i expected, heh.
I think JSON is a fine method, but everything would need to be broken up into chunks in separate files, and there should be a file structure that mimics the DB structure in a way. Then it can be ZIP'd to take up a lot less storage.

**Reply by TaeHagen** [reply]
Yeah tbh I was kinda lazy and just wanted something quick and easy that was good enough for now. If json size without attachments isn't an issue we could just gzip that current file to save storage.
